### ISSUE TITLE
Add __hash__ function to avoid unnecessary recompilations

### ISSUE THREAD

**Issue by dionhaefner** [deepmind/optax#353]
Example:

```python
from functools import partial
import jax
import jax.numpy as jnp
import optax

@partial(jax.jit, static_argnames=("optimizer",))
def train_step(optimizer, opt_state):
    print("compiling")

def train():
    optimizer = optax.adam(1e-2)
    opt_state = optimizer.init({"foo": jnp.zeros((100, 100))})
    
    for _ in range(10):
        train_step(optimizer, opt_state)

train()
train()
train()
``` 

This prints:

```
compiling
compiling
compiling
```

Even though the same optimizer is used every time. This could be avoided by supplying meaningful `__hash__` attributes based on the optimizer parameters, or memoizing the wrappers in `alias.py`.

**Comment by PhilipVinc** [comment]
or even better, by marking optimisers as pytrees and splitting 'dynamic' data such as learning rate and such from static data...

**Comment by NeilGirdhar** [comment]
> or even better, by marking optimisers as pytrees and splitting 'dynamic' data such as learning rate and such from static data...

Exactly!  This is the approach I took in my [shim library](https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/gradient/transforms.py) in which I just call through to Optax.

**Comment by natinew77-creator** [comment]
Hi! I'm interested in contributing to this issue.

After reviewing the discussion, I see two potential approaches:

1. **Simple `__hash__`/`__eq__` approach**: Add hash methods based on optimizer parameters to `GradientTransformation`. This would be a minimal change that solves the immediate recompilation issue.

2. **PyTree-based approach** (as suggested by @PhilipVinc and @NeilGirdhar): Register optimizers as JAX pytrees, splitting dynamic data (learning rate, etc.) from static data. This is more elegant and would also prevent recompilation when only dynamic parameters change.

I've looked at @NeilGirdhar's [tjax implementation](https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/gradient/transforms.py) as a reference.

**Questions for maintainers:**
- Would a simpler `__hash__`-based solution be acceptable as a first step, or would you prefer the full pytree approach?
- Are there any specific optimizers that should be prioritized?

Happy to start working on a PR if there's interest!

**Comment by NeilGirdhar** [comment]
Thanks for looking at this with fresh eyes!

> 1. Add hash methods based on optimizer parameters to `GradientTransformation`. This would be a minimal change that solves the immediate recompilation issue.

Just FYI: While that does solve the problem in the issue, that won't solve the problem in general.  If the parameters are changed, the hash must change, and the jit will want to recompile when the hash changes.  If, instead, the parameters were dynamic, then no recompilation would be necessary for parameter changes.

**Comment by natinew77-creator** [comment]
Thanks for the clear explanation, @NeilGirdhar! That makes perfect sense treating parameters as dynamic to avoid recompilation upon updates is definitely the more robust solution.

I will proceed with Option 2 (the PyTree approach). I'll start by prototyping this change for a single optimizer (likely Adam) to demonstrate the split between static and dynamic data, and then report back with the results.

I'll get started on that now!
